pkg/attestation/crafter/materials/oci_image.go

```diff
@@ -60,13 +81,119 @@ func (i *OCIImageCrafter) Craft(_ context.Context, imageRef string) (*api.Attest
 	repoName := remoteRef.Repository.String()
 	i.logger.Debug().Str("name", repoName).Str("digest", remoteRef.DigestStr()).Msg("container image resolved")
 
+	// Check if the signature tag exists for the image digest
+	signatureInfo := i.checkForSignature(ref, descriptor)
+
+	containerImage := &api.Attestation_Material_ContainerImage{
+		Id: i.input.Name, Name: repoName, Digest: remoteRef.DigestStr(), IsSubject: i.input.Output,
+		Tag: ref.Identifier(),
+	}
+
+	// If the signature digest exists, add it to the material
+	if signatureInfo != nil {
+		containerImage.SignatureDigest = signatureInfo.digest
+		containerImage.SignatureProvider = string(signatureInfo.provider)
+	}
+
 	return &api.Attestation_Material{
 		MaterialType: i.input.Type,
-		M: &api.Attestation_Material_ContainerImage_{
-			ContainerImage: &api.Attestation_Material_ContainerImage{
-				Id: i.input.Name, Name: repoName, Digest: remoteRef.DigestStr(), IsSubject: i.input.Output,
-				Tag: ref.Identifier(),
-			},
-		},
+		M:            &api.Attestation_Material_ContainerImage_{ContainerImage: containerImage},
 	}, nil
 }
+
+// checkForSignature checks for a signature for the given image reference.
+// Right now it checks for Cosign signatures only by tag or referrers API.
+func (i *OCIImageCrafter) checkForSignature(originalRef name.Reference, originalDesc *remote.Descriptor) *ContainerSignatureInfo {
+	var wg sync.WaitGroup
+	resultChan := make(chan *ContainerSignatureInfo, 1)
+
+	// Launch parallel goroutines for signature checks
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		if sigInfo := i.checkForSignatureTag(originalRef, originalDesc); sigInfo != nil {
+			select {
+			case resultChan <- sigInfo:
+			default:
+				// Channel already closed, result obtained from another source
+			}
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		if sigInfo := i.checkForSignatureReferrers(originalRef, originalDesc); sigInfo != nil {
+			select {
+			case resultChan <- sigInfo:
+			default:
+				// Channel already closed, result obtained from another source
+			}
+		}
+	}()
+
+	go func() {
+		wg.Wait()
```

Comment:
what value do we get of doing this wait async? Shouodn't it be done outside of the goroutine?

Reply:
It can be done outside the goroutine but in this way we make explicit that we are waiting on the channel to get that and block there. In this scenario is the same